Feature/us 022 imp add validations to post operation to create users

RamOwen99 opened the pull request:
# Validations to post operation to create users
In this Pull Request, I did the validation on the data that the user gave us to create a new user. The validations before storing the user's data in the database:
- The **type** field should be only 1  (admin) or 2 (normal user)
- The fields **first name**, **last name**, **email**, **password** should be required
- The field **email** needs to be a validated email (@ + domain + extension)
- The field **password**  needs to have the correct format.:
    - At least 10 characters
    - At least 1 numeric value
-  The **email** that the user gave us needs to be unique in all the database
- The fields **first name** and **last name** will store in capital letters

RamOwen99 committed:
Update pom.xml

RamOwen99 committed:
Update Validator interface

Delete blank spaces on the file, and delete an unnecessary import and the
public word as well

RamOwen99 committed:
Add new validation methods

Add these methods:
	- validate if a string doesn't have content, if is null, or if
	its content it is only blank spaces
	- validate if a string is lowercase
	- validate the password format. In this project the format will
	  be at least 10 characters and it needs to have one numeric
	  value
Also, update the regular expression on the email

RamOwen99 committed:
Create method to know if an email exist in the db

RamOwen99 committed:
Handle a bad request with a message

RamOwen99 committed:
Validate user's data

Before to store the data, we need to check if all the data that the user
gave us it is correct, if it is correct we store the data, but if it
is not, we display a message of what could be wrong

ivanflores15 commented on pom.xml:
```diff
@@ -78,7 +78,15 @@
 				</configuration>
 			</plugin>
-		</plugins>
+            <plugin>
```
is this plugin used by implementation? 
if yes, Heroku containers just accept Java 8, so probably we will run into a problem because of version.

RamOwen99 replied:
No, it just updates when I clone the repository

ivanflores15 replied:
So, is it necessary or isn't?

ivanflores15 replied:
about this plugin, is still necessary?

ivanflores15 commented on src/main/java/com/agilethought/atsceapi/exception/ErrorMessage.java:
```diff
@@ -0,0 +1,5 @@
+package com.agilethought.atsceapi.exception;
+
+public class ErrorMessage {
+    public static final String badRequestMessage = "You cannot obtain the ";
```
"You cannot obtain the " I don't think this is a good message for `BadRequests messages`. @EmiliaGuillen-dev and @luisgalindoluna what do you think?

luisgalindoluna replied:
Yeah, try to be more specific with these error messages. That sentence doesn't really tell us a lot.

ivanflores15 commented on src/main/java/com/agilethought/atsceapi/repository/UserRepository.java:
```diff
@@ -14,4 +14,6 @@
 	@Query("{ 'email' : ?0, 'password' : ?1 }")
 	List<User> findUsersByEmail(String email, String password);
+
+	boolean existsByEmail(String email);
```
What is reason to use a primitive `boolean`?

ivanflores15 commented on src/main/java/com/agilethought/atsceapi/service/UserServiceImpl.java:
```diff
@@ -14,7 +15,6 @@
 import com.agilethought.atsceapi.exception.UnauthorizedException;
 import com.agilethought.atsceapi.model.User;
 import com.agilethought.atsceapi.repository.UserRepository;
-import com.agilethought.atsceapi.validator.LoginValidator;
```
LoginValidator was removed but I see a dependency in line 34.

RamOwen99 replied:
It's because we inject that dependency to do the post operation for login a user

ivanflores15 commented on src/main/java/com/agilethought/atsceapi/validator/UserValidator.java:
```diff
@@ -0,0 +1,67 @@
+
+    private void validateUserTypeField(int type) {
+        if(type < 1 || type > 2) {
+            log.info("fail type");
```
I think logs must be more informative. @luisgalindoluna @EmiliaGuillen-dev what do you think?

ivanflores15 commented on src/main/java/com/agilethought/atsceapi/validator/UserValidator.java:
```diff
@@ -0,0 +1,67 @@
+        validateUserLastName(user.getLastName());
+        validateEmailFormat(user.getEmail());
+        validatePasswordFormat(user.getPassword());
+        validateEmailInTheDataBase(user.getEmail());
```
`validateEmailInTheDataBase` article The is unnecesarry.

ivanflores15 commented on src/main/java/com/agilethought/atsceapi/validator/UserValidator.java:
```diff
@@ -0,0 +1,67 @@
+        validateEmailFormat(user.getEmail());
+        validatePasswordFormat(user.getPassword());
+        validateEmailInTheDataBase(user.getEmail());
+        user.setFirstName(user.getFirstName().toUpperCase());
```
Why are you setting User properties inside of Validator Class?. It doesn't make sense with Class purposes

ivanflores15 commented on src/main/java/com/agilethought/atsceapi/validator/ValidationUtils.java:
```diff
@@ -1,22 +1,34 @@
+		return matcherEmail.find();
+	}
+
+	static boolean isValidatePassword(String password) {
```
Validate is a verb, it doesn't make sense with prior verb"is". you must use an adjective after "is" verb.

ivanflores15 reviewed (commented):
We normally left a space between class declaration and body of class itself, example: 
`public class SomeClass { `
` // space `
` @Autowired
private BeanClass injection
`
This is just a recommendation.

ivanflores15 commented on src/main/java/com/agilethought/atsceapi/validator/UserValidator.java:
```diff
@@ -0,0 +1,67 @@
+
+    private void validateEmailFormat(String email) {
+        if(!isStringValid(email) || !isStringLowerCase(email) || !isValidEmail(email)) {
+            throw new BadRequestException(badRequestMessage + "*email* because it doesn't have the correct format");
```
`because it doesn't have the correct format` you are repeting this String 3 times, don't you think is a good idea to have a constant to replace it?

luisgalindoluna reviewed (changes requested):
Also notice that your pom.xml file has conflicts.

RamOwen99 committed:
Update error message

To have a better description when the program fail, the messages now are
more descriptive.

On the ValidatonUtils, I modify the name of the method that validate if
the password has the correct format

RamOwen99 committed:
Set first and last name in capital letters

As a requirement, the user's firstName and lastName needs to store on
the database in uppercase

ivanflores15 commented on src/main/java/com/agilethought/atsceapi/exception/ErrorMessage.java:
```diff
@@ -0,0 +1,10 @@
+
+public class ErrorMessage {
+
+    public static final String badRequestMessage = "The field ";
```
All contans name must be named in UpperCase separated by _. Example: `BAD_REQUEST_MESSAGE`

ivanflores15 commented on src/main/java/com/agilethought/atsceapi/service/UserServiceImpl.java:
```diff
@@ -74,8 +77,10 @@ public void deleteUserById(String id) {
 	public NewUserResponse createUser(NewUserRequest request) {
 		User user = orikaMapperFacade.map(request, User.class);
+		userValidator.validate(user);
+		user.setFirstName(user.getFirstName().toUpperCase());
```
I couldn't find this requirement into your US, does it really exist?

ivanflores15 commented on src/main/java/com/agilethought/atsceapi/validator/UserValidator.java:
```diff
@@ -0,0 +1,65 @@
+import static com.agilethought.atsceapi.validator.ValidationUtils.*;
+
+@Service
+@Slf4j
```
you never used logs or something related to

RamOwen99 committed:
Merge branch 'develop' into feature/US_022_Imp_Add_Validations_To_Post_Operation_To_Create_Users

RamOwen99 committed:
Rename constants message

ivanflores15 reviewed (approved):
looks good

luisgalindoluna reviewed (approved):
Looks good

EmiliaGuillen-dev reviewed (approved):
looks good
